Sources/MQTTNIO/ChannelHandlers/MQTTChannelHandler.swift

```diff
@@ -202,91 +202,53 @@ class MQTTChannelHandler: ChannelDuplexHandler {
     // MARK: - Sending Messages
 
     private func sendMessage(
-        channel: any Channel,
         _ message: MQTTPacket,
         promise: MQTTPromise<MQTTPacket>,
         checkInbound: @escaping (MQTTPacket) throws -> Bool
     ) {
         let task = MQTTTask(
             promise: promise,
-            on: channel.eventLoop,
+            on: self.eventLoop,
             timeout: self.configuration.timeout,
             checkInbound: checkInbound
         )
 
-        self.addTask(task)
-            .flatMap {
-                channel.writeAndFlush(message)
-            }
-            .whenFailure { error in
-                task.fail(error)
+        switch self.stateMachine.sendPacket(task) {
```

Comment:
You need to be sure any calls to the stateMachine are done on the EventLoop. There is a good chance the `sendMessage` at this point will not be called on the event loop. I would add an assert here `self.eventLoop.assertInEventLoop()` and then add in `MQTTConnection` checks for `inEventLoop` etc. You can also remove the check for `inEventLoop` further down in this function if you are certain you are running on the event loop.